lib/src/rules/collection_methods_unrelated_type.dart

```diff
@@ -0,0 +1,229 @@
+}
+
+abstract class ClassBase {}
+
+class DerivedClass1 extends ClassBase {}
+```
+
+**GOOD:**
+```dart
+void someFunction6() {
+  List<Mixin> list = <Mixin>[];
+  DerivedClass2 instance;
+  if (list.contains(instance)) print('someFunction6'); // OK
+}
+
+abstract class ClassBase {}
+
+abstract class Mixin {}
+
+class DerivedClass2 extends ClassBase with Mixin {}
+```
+
+**GOOD:**
+```dart
+void someFunction7() {
+  List<Mixin> list = <Mixin>[];
+  DerivedClass3 instance;
+  if (list.contains(instance)) print('someFunction7'); // OK
+}
+
+abstract class ClassBase {}
+
+abstract class Mixin {}
+
+class DerivedClass3 extends ClassBase implements Mixin {}
+```
+
+''';
+
+class CollectionMethodsUnrelatedType extends LintRule {
+  static const LintCode code = LintCode('collection_methods_unrelated_type',
+      "The type of the argument of 'Iterable<{0}>.contains' isn't a subtype of '{0}'.");
+
+  CollectionMethodsUnrelatedType()
+      : super(
+            name: 'collection_methods_unrelated_type',
+            description: _desc,
+            details: _details,
+            group: Group.errors);
+
+  @override
+  LintCode get lintCode => code;
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this, context.typeSystem, context.typeProvider);
+    registry.addMethodInvocation(this, visitor);
+  }
+}
+
+class _Visitor extends UnrelatedTypesProcessors {
+  _Visitor(super.rule, super.typeSystem, super.typeProvider);
+
+  @override
+  List<MethodDefinition> get methods => [
+        // Argument to `Iterable<E>.contains` should be assignable to `E`.
+        MethodDefinition(
+          typeProvider.iterableElement,
+          'contains',
+          ExpectedArgumentKind.assignableToCollectionTypeArgument,
+        ),
+        // Argument to `List<E>.remove` should be assignable to `E`.
+        MethodDefinition(
+          typeProvider.listElement,
+          'remove',
+          ExpectedArgumentKind.assignableToCollectionTypeArgument,
+        ),
+        // Argument to `Map<K, V>.containsKey` should be assignable to `K`.
+        MethodDefinition(
+          typeProvider.mapElement,
+          'containsKey',
+          ExpectedArgumentKind.assignableToCollectionTypeArgument,
+        ),
+        // Argument to `Map<K, V>.containsValue` should be assignable to `V`.
+        MethodDefinition(
+          typeProvider.mapElement,
+          'containsValue',
+          ExpectedArgumentKind.assignableToCollectionTypeArgument,
+          typeArgumentIndex: 1,
+        ),
+        // Argument to `Map<K, V>.remove` should be assignable to `K`.
+        MethodDefinition(
+          typeProvider.mapElement,
+          'remove',
+          ExpectedArgumentKind.assignableToCollectionTypeArgument,
+        ),
+        // TODO(srawlins): Queue? It's not in mock SDK or [TypeProvider].
+        // Argument to `Queue<E>.remove` should be assignable to `E`.
```

Comment:
We only have accessors in `TypeProvide` for the types from the SDK that are referenced in the spec (because those are the ones that `analyzer` needs to use to validate the code. You'll need to add it to the mock sdk and then use something like `typeProvider.objectElement.library.getClass('Queue')` to get the element.